### PR TITLE
distrobox: 1.5.0.2 -> 1.6.0.1

### PR DESCRIPTION
**Pull request by pongo1231**
## Description of changes

https://github.com/89luca89/distrobox/releases/tag/1.6.0.1

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/virtualization/distrobox/default.nix [hunk 1]
```diff
@@ -2,13 +2,13 @@
 
 stdenvNoCC.mkDerivation (finalAttrs: {
   pname = "distrobox";
-  version = "1.5.0.2";
+  version = "1.6.0";
```
Comment (pschmitt) [inline pkgs/applications/virtualization/distrobox/default.nix:5]: Version 1.6.0.1 has been released:
```suggestion
  version = "1.6.0.1";
```

Comment (pschmitt) [reply pkgs/applications/virtualization/distrobox/default.nix:5]: https://github.com/89luca89/distrobox/releases/tag/1.6.0.1

File: pkgs/applications/virtualization/distrobox/default.nix [hunk 2]
```diff
@@ -2,13 +2,13 @@
...
 
   src = fetchFromGitHub {
     owner = "89luca89";
     repo = finalAttrs.pname;
     rev = finalAttrs.version;
-    hash = "sha256-ss8049D6n1V/gDzEMjywDnoke5s2we9j3mO8yta72UA=";
+    hash = "sha256-AwsRZS296laGFDlV/xGBuWKXuOP2qB+MKGE471BbPt8=";
```
Comment (pschmitt) [inline pkgs/applications/virtualization/distrobox/default.nix:11]:
```suggestion
    hash = "sha256-UWrXpb20IHcwadPpwbhSjvOP1MBXic5ay+nP+OEVQE4=";
```

**Comment by pongo1231** [comment]
Just tried to create a container
```
$ distrobox create --name test --image debian
Creating 'test' using image debian       [ OK ]
Distrobox 'test' successfully created.
To enter, run:

distrobox enter test

$ distrobox enter test
 Error: could not start entrypoint.
{"msg":"exec container process (missing dynamic library?) `/usr/bin/entrypoint`: No such file or directory","level":"error","time":"2023-11-22T00:15:15.341593Z"}
```

**Comment by pschmitt** [comment]
Yeah I noticed this as well. My best guess: the issue is that `distrobox-init` is a symlink.

**Comment by pschmitt** [comment]
https://github.com/89luca89/distrobox/blob/3435f4d27070a99668bfa29a3e508db4ecc09009/distrobox-create#L77-L80

**Comment by pongo1231** [comment]
Been able to pinpoint the regression to https://github.com/89luca89/distrobox/commit/5dd7b00e3f095e1976c0e2aa4f972efdcd6f4f3a. Not sure why that is though as I don't think those mounts should matter?

> I don't get the rationale behind that comment, i believe resolving the symlink is what we actually want here.

So it always mounts `/run/current-system/sw/bin/distrobox-init` as `/usr/bin/entrypoint` in the container rather than a specific store path which might eventually get garbage collected, no?

**Comment by pschmitt** [comment]
> Been able to pinpoint the regression to https://github.com/89luca89/distrobox/commit/5dd7b00e3f095e1976c0e2aa4f972efdcd6f4f3a. Not sure why that is though as I don't think those mounts should matter?

My guess is that there are just too many symlinks in the entire chain here:

```shell
» where distrobox-init
/etc/profiles/per-user/pschmitt/bin/distrobox-init
» ls -l /etc/profiles/per-user/pschmitt/bin/distrobox-init
lrwxrwxrwx 2 root root 80  1. Jan 1970  /etc/profiles/per-user/pschmitt/bin/distrobox-init -> /nix/store/gwmsb1lpc1a0jvb2dc6yf7g27lv0662a-home-manager-path/bin/distrobox-init
» ls -l /nix/store/gwmsb1lpc1a0jvb2dc6yf7g27lv0662a-home-manager-path/bin/distrobox-init
lrwxrwxrwx 2 root root 80  1. Jan 1970  /nix/store/gwmsb1lpc1a0jvb2dc6yf7g27lv0662a-home-manager-path/bin/distrobox-init -> /nix/store/cci7i1s0pfdnpwva68plrbspcnza37bv-distrobox-1.6.0.1/bin/distrobox-init
» ls -l /nix/store/cci7i1s0pfdnpwva68plrbspcnza37bv-distrobox-1.6.0.1/bin/distrobox-init
-r-xr-xr-x 20 root root 64968  1. Jan 1970  /nix/store/cci7i1s0pfdnpwva68plrbspcnza37bv-distrobox-1.6.0.1/bin/distrobox-init
```

The CRI might not handle this all too well, as in if the symlink does get bind-mounted directly into the container that would explain why it not having access to /nix makes this fail. 

> So it always mounts /run/current-system/sw/bin/distrobox-init as /usr/bin/entrypoint in the container rather than a specific store path which might eventually get garbage collected, no?

These paths wouldn't get gc'd away since the package would still be installed.


The following patch makes 1.6.0.1 usable:

```patch
diff --git a/distrobox-create b/distrobox-create
index 64e9911..21c707b 100755
--- a/distrobox-create
+++ b/distrobox-create
@@ -74,14 +74,15 @@ unshare_ipc=0
 unshare_netns=0
 unshare_process=0
 unshare_devsys=0
+
+self_dir="$(dirname "$(realpath "$0")")"
+nix_config_file="$(realpath "${self_dir}/../share/distrobox/distrobox.conf")"
+
 # Use cd + dirname + pwd so that we do not have relative paths in mount points
-# We're not using "realpath" here so that symlinks are not resolved this way
-# "realpath" would break situations like Nix or similar symlink based package
-# management.
-distrobox_entrypoint_path="$(cd "$(dirname "${0}")" && pwd)/distrobox-init"
-distrobox_export_path="$(cd "$(dirname "${0}")" && pwd)/distrobox-export"
-distrobox_genentry_path="$(cd "$(dirname "${0}")" && pwd)/distrobox-generate-entry"
-distrobox_hostexec_path="$(cd "$(dirname "${0}")" && pwd)/distrobox-host-exec"
+distrobox_entrypoint_path="${self_dir}/distrobox-init"
+distrobox_export_path="${self_dir}/distrobox-export"
+distrobox_genentry_path="${self_dir}/distrobox-generate-entry"
+distrobox_hostexec_path="${self_dir}/distrobox-host-exec"
 # In case some of the scripts are not in the same path as create, let's search
 # in PATH for them.
 [ ! -e "${distrobox_entrypoint_path}" ] && distrobox_entrypoint_path="$(command -v distrobox-init)"
@@ -98,6 +99,7 @@ version="1.6.0.1"
 # priority over system defaults
 # leave priority to environment variables.
 config_files="
+	$nix_config_file
 	/usr/share/distrobox/distrobox.conf
 	/usr/share/defaults/distrobox/distrobox.conf
 	/usr/etc/distrobox/distrobox.conf
diff --git a/distrobox-enter b/distrobox-enter
index 6d8998a..bb05437 100755
--- a/distrobox-enter
+++ b/distrobox-enter
@@ -75,10 +75,14 @@ skip_workdir=0
 verbose=0
 version="1.6.0.1"
 
+self_dir="$(dirname "$(realpath "$0")")"
+nix_config_file="$(realpath "${self_dir}/../share/distrobox/distrobox.conf")"
+
 # Source configuration files, this is done in an hierarchy so local files have
 # priority over system defaults
 # leave priority to environment variables.
 config_files="
+	$nix_config_file
 	/usr/share/distrobox/distrobox.conf
 	/usr/share/defaults/distrobox/distrobox.conf
 	/usr/etc/distrobox/distrobox.conf
diff --git a/distrobox-generate-entry b/distrobox-generate-entry
index 3243691..6a7910a 100755
--- a/distrobox-generate-entry
+++ b/distrobox-generate-entry
@@ -45,10 +45,14 @@ icon_default="${HOME}/.local/share/icons/terminal-distrobox-icon.svg"
 verbose=0
 version="1.6.0.1"
 
+self_dir="$(dirname "$(realpath "$0")")"
+nix_config_file="$(realpath "${self_dir}/../share/distrobox/distrobox.conf")"
+
 # Source configuration files, this is done in an hierarchy so local files have
 # priority over system defaults
 # leave priority to environment variables.
 config_files="
+	$nix_config_file
 	/usr/share/distrobox/distrobox.conf
 	/usr/share/defaults/distrobox/distrobox.conf
 	/usr/etc/distrobox/distrobox.conf
diff --git a/distrobox-list b/distrobox-list
index aaec85e..235e529 100755
--- a/distrobox-list
+++ b/distrobox-list
@@ -44,10 +44,14 @@ verbose=0
 version="1.6.0.1"
 container_manager="autodetect"
 
+self_dir="$(dirname "$(realpath "$0")")"
+nix_config_file="$(realpath "${self_dir}/../share/distrobox/distrobox.conf")"
+
 # Source configuration files, this is done in an hierarchy so local files have
 # priority over system defaults
 # leave priority to environment variables.
 config_files="
+	$nix_config_file
 	/usr/share/distrobox/distrobox.conf
 	/usr/share/defaults/distrobox/distrobox.conf
 	/usr/etc/distrobox/distrobox.conf
diff --git a/distrobox-rm b/distrobox-rm
index 702c1dd..2e37538 100755
--- a/distrobox-rm
+++ b/distrobox-rm
@@ -54,10 +54,14 @@ rm_home=0
 response_rm_home="N"
 version="1.6.0.1"
 
+self_dir="$(dirname "$(realpath "$0")")"
+nix_config_file="$(realpath "${self_dir}/../share/distrobox/distrobox.conf")"
+
 # Source configuration files, this is done in an hierarchy so local files have
 # priority over system defaults
 # leave priority to environment variables.
 config_files="
+	$nix_config_file
 	/usr/share/distrobox/distrobox.conf
 	/usr/share/defaults/distrobox/distrobox.conf
 	/usr/etc/distrobox/distrobox.conf
diff --git a/distrobox-stop b/distrobox-stop
index fd17cc1..e0dbc8f 100755
--- a/distrobox-stop
+++ b/distrobox-stop
@@ -52,10 +52,14 @@ non_interactive=0
 verbose=0
 version="1.6.0.1"
 
+self_dir="$(dirname "$(realpath "$0")")"
+nix_config_file="$(realpath "${self_dir}/../share/distrobox/distrobox.conf")"
+
 # Source configuration files, this is done in an hierarchy so local files have
 # priority over system defaults
 # leave priority to environment variables.
 config_files="
+	$nix_config_file
 	/usr/share/distrobox/distrobox.conf
 	/usr/share/defaults/distrobox/distrobox.conf
 	/usr/etc/distrobox/distrobox.conf
diff --git a/distrobox-upgrade b/distrobox-upgrade
index ab5e96f..dc8d295 100755
--- a/distrobox-upgrade
+++ b/distrobox-upgrade
@@ -39,10 +39,14 @@ rootful=0
 verbose=0
 version="1.6.0.1"
 
+self_dir="$(dirname "$(realpath "$0")")"
+nix_config_file="$(realpath "${self_dir}/../share/distrobox/distrobox.conf")"
+
 # Source configuration files, this is done in an hierarchy so local files have
 # priority over system defaults
 # leave priority to environment variables.
 config_files="
+	$nix_config_file
 	/usr/share/distrobox/distrobox.conf
 	/usr/share/defaults/distrobox/distrobox.conf
 	/usr/etc/distrobox/distrobox.conf
```

```nix
{ stdenvNoCC, lib, fetchFromGitHub, makeWrapper, wget }:

stdenvNoCC.mkDerivation (finalAttrs: {
  pname = "distrobox";
  version = "1.6.0.1";

  src = fetchFromGitHub {
    owner = "89luca89";
    repo = finalAttrs.pname;
    rev = finalAttrs.version;
    hash = "sha256-UWrXpb20IHcwadPpwbhSjvOP1MBXic5ay+nP+OEVQE4=";
  };

  dontConfigure = true;
  dontBuild = true;

  nativeBuildInputs = [ makeWrapper ];

  patches = [ ./resolve-symlinks.patch ];

  installPhase = ''
    runHook preInstall

    # https://github.com/89luca89/distrobox/issues/408
    substituteInPlace ./distrobox-generate-entry \
      --replace 'icon_default="''${HOME}/.local' "icon_default=\"$out"
    ./install -P $out

    runHook postInstall
  '';

  # https://github.com/89luca89/distrobox/issues/407
  postFixup = ''
    wrapProgram "$out/bin/distrobox-generate-entry" \
      --prefix PATH ":" ${lib.makeBinPath [ wget ]}

    mkdir -p $out/share/distrobox
    echo 'container_additional_volumes="/nix:/nix"' > $out/share/distrobox/distrobox.conf
  '';

  meta = with lib; {
    description = "Wrapper around podman or docker to create and start containers";
    longDescription = ''
      Use any linux distribution inside your terminal. Enable both backward and
      forward compatibility with software and freedom to use whatever distribution
      you’re more comfortable with
    '';
    homepage = "https://distrobox.privatedns.org/";
    license = licenses.gpl3Only;
    platforms = platforms.linux;
    maintainers = with maintainers; [ atila ];
  };
})
```

**Comment by pschmitt** [comment]
The quickest win for the nixpkg might be to ship a static config that enables mouting of /nix in all the containers.

**Comment by pongo1231** [comment]
Appears to work fine from brief testing after applying your changes. Thanks!

File: pkgs/applications/virtualization/distrobox/default.nix [hunk 3]
```diff
@@ -2,19 +2,22 @@
...
-    hash = "sha256-ss8049D6n1V/gDzEMjywDnoke5s2we9j3mO8yta72UA=";
+    hash = "sha256-UWrXpb20IHcwadPpwbhSjvOP1MBXic5ay+nP+OEVQE4=";
   };
 
   dontConfigure = true;
   dontBuild = true;
 
   nativeBuildInputs = [ makeWrapper ];
+
+  patches = [ ./always-mount-nix.patch ];
```
Comment (Mindavi) [inline pkgs/applications/virtualization/distrobox/default.nix:19]: It's a good idea to reference the upstream patch here, to make the tracking easier.

**Review by Mindavi — APPROVED** [review]

File: pkgs/applications/virtualization/distrobox/default.nix (hunk 3 above)
Comment (pongo1231) [reply pkgs/applications/virtualization/distrobox/default.nix:20]: Added a reference to the upstream PR

**Comment by Aleksanaa** [comment]
Please also change `meta.homepage` from https://distrobox.privatedns.org/ to https://distrobox.it/ . Currently the former is a redirect page.

**Comment by pongo1231** [comment]
Replaced `finalAttrs.pname` with `distrobox` for the repo name as per [this suggestion](https://github.com/NixOS/nixpkgs/pull/275684#discussion_r1433346616) from the other PR.

**Comment by Mindavi** [comment]
Fell out of sight for me, but yeah, let's do this

**Pull request merged**